core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -124,162 +136,170 @@ static Builder buildFrom(ViewMetadata base) {
+      changes.add(new MetadataUpdate.SetLocation(newLocation));
       return this;
     }
 
     public Builder setCurrentVersionId(int newVersionId) {
-      if (newVersionId == -1) {
+      if (newVersionId == LAST_ADDED) {
         ValidationException.check(
             lastAddedVersionId != null,
             "Cannot set last version id: no current version id has been set");
         return setCurrentVersionId(lastAddedVersionId);
       }
 
-      if (this.currentVersionId == newVersionId) {
+      if (currentVersionId == newVersionId) {
         return this;
       }
 
-      checkCurrentVersionIdIsValid(newVersionId);
-
-      // by that time, schemas are defined and the highestFieldId can be determined and schema
-      // changes can be added
-      int highestFieldId = highestFieldId();
-      for (Schema schema : schemas) {
-        this.changes.add(new MetadataUpdate.AddSchema(schema, highestFieldId));
-      }
+      ViewVersion version = versionsById.get(newVersionId);
+      Preconditions.checkArgument(
+          version != null, "Cannot set current version to unknown version: %s", newVersionId);
 
       this.currentVersionId = newVersionId;
 
       if (lastAddedVersionId != null && lastAddedVersionId == newVersionId) {
-        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(LAST_ADDED));
+        changes.add(new MetadataUpdate.SetCurrentViewVersion(LAST_ADDED));
       } else {
-        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+        changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
       }
 
       return this;
     }
 
-    private int highestFieldId() {
-      return schemas.stream().map(Schema::highestFieldId).max(Integer::compareTo).orElse(0);
-    }
-
-    public Builder setCurrentVersion(ViewVersion version) {
-      Schema schema = indexSchemas(schemas).get(version.schemaId());
+    public Builder setCurrentVersion(ViewVersion version, Schema schema) {
       int newSchemaId = addSchemaInternal(schema);
-      return setCurrentVersionId(
-          addVersionInternal(
-              ImmutableViewVersion.builder().from(version).schemaId(newSchemaId).build()));
+      ViewVersion newVersion =
+          ImmutableViewVersion.builder().from(version).schemaId(newSchemaId).build();
+      return setCurrentVersionId(addVersionInternal(newVersion));
     }
 
-    private int reuseOrCreateNewViewVersionId(ViewVersion viewVersion) {
-      // if the view version already exists, use its id; otherwise use the highest id + 1
-      int newVersionId = viewVersion.versionId();
-      for (ViewVersion version : versions) {
-        if (version.equals(viewVersion)) {
-          return version.versionId();
-        } else if (version.versionId() >= newVersionId) {
-          newVersionId = viewVersion.versionId() + 1;
-        }
-      }
-
-      return newVersionId;
+    public Builder addVersion(ViewVersion version) {
```

Comment:
Moved some methods around to keep version and schema methods located near one another.